File: resources/generators/lib/schema_validate.py
"""CLI bridge to opengwasdb's shared `analyses.tsv` schema (issue #51).

R generator scripts have no Python import machinery, so this is a thin CLI
wrapper around `opengwasdb.model.analyses.validate_analyses()` they can shell
out to, rather than hand-maintaining a duplicate required-column/vocabulary
list in R that can silently drift from opengwasdb's own (the problem issue
#51 exists to close). All interpretation of what's required and which
vocabularies apply lives in opengwasdb; this script only reads a manifest and
reports what opengwasdb says about it.

Usage:

    python3 resources/generators/lib/schema_validate.py <analyses.tsv path>

Exits 0 with no output when the file conforms to the shared core schema.
Exits 1 with one violation per stdout line otherwise.
"""
from __future__ import annotations

import argparse
import sys

from opengwasdb.model.analyses import read_analyses, validate_analyses


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("analyses_path", help="Path to an emitted analyses.tsv")
    args = parser.parse_args(argv)

    table = read_analyses(args.analyses_path)
    errors = validate_analyses(table)
    for error in errors:
        print(error)
    return 1 if errors else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: resources/generators/lib/schema_validate.R
# R bridge to opengwasdb's shared analyses.tsv schema (issue #51). Generator
# scripts shell out to resources/generators/lib/schema_validate.py rather than hand-
# maintaining a duplicate required-column/vocabulary list in R that can
# silently drift from what opengwasdb itself considers valid -- the failure
# mode this repository and opengwasdb independently defining analyses.tsv
# schemas invites (see docs/adr/0017-opengwasdb-owns-shared-analysis-schema.md).

validate_against_opengwasdb_schema <- function(analyses_path, root) {
  python_bin <- Sys.which("python3")
  if (!nzchar(python_bin)) {
    stop("python3 not found on PATH; required to validate analyses.tsv against opengwasdb's shared schema")
  }
  script <- normalizePath(file.path(root, "resources/generators/lib/schema_validate.py"), winslash = "/", mustWork = FALSE)
  output <- suppressWarnings(system2(
    python_bin, c(shQuote(script), shQuote(analyses_path)),
    stdout = TRUE, stderr = TRUE
  ))
  # system2(..., stdout=TRUE) only attaches a "status" attribute when the
  # child process exits non-zero; a clean pass leaves it NULL.
  exit_status <- attr(output, "status")
  passed <- is.null(exit_status) || identical(as.integer(exit_status), 0L)
  list(passed = passed, errors = if (passed) character() else output)
}
